client/service/src/builder.rs

```diff
@@ -690,55 +674,53 @@ async fn transaction_notifications<TBl, TExPool>(
 		.await;
 }
 
-fn build_telemetry<TBl: BlockT>(
+fn init_telemetry<TBl: BlockT, TCl: BlockBackend<TBl>>(
 	config: &mut Configuration,
-	endpoints: sc_telemetry::TelemetryEndpoints,
-	telemetry_connection_sinks: TelemetryConnectionSinks,
 	network: Arc<NetworkService<TBl, <TBl as BlockT>::Hash>>,
-	spawn_handle: SpawnTaskHandle,
-	genesis_hash: <TBl as BlockT>::Hash,
-) -> sc_telemetry::Telemetry {
+	client: Arc<TCl>,
+) -> Option<sc_telemetry::TelemetryConnectionNotifier> {
+	let telemetry_handle = if let Some(handle) = config.telemetry_handle.as_mut() {
+		handle
+	} else {
+		return None;
+	};
+
+	let endpoints = match config.telemetry_endpoints.clone() {
+		// Don't initialise telemetry if `telemetry_endpoints` == Some([])
+		Some(endpoints) if !endpoints.is_empty() => endpoints,
+		_ => return None,
+	};
+
+	let genesis_hash = match client.block_hash(Zero::zero()) {
+		Ok(Some(hash)) => hash,
+		_ => Default::default(),
+	};
+
 	let is_authority = config.role.is_authority();
 	let network_id = network.local_peer_id().to_base58();
 	let name = config.network.node_name.clone();
 	let impl_name = config.impl_name.clone();
 	let impl_version = config.impl_version.clone();
 	let chain_name = config.chain_spec.name().to_owned();
-	let telemetry = sc_telemetry::init_telemetry(sc_telemetry::TelemetryConfig {
-		endpoints,
-		wasm_external_transport: config.telemetry_external_transport.take(),
-	});
 	let startup_time = SystemTime::UNIX_EPOCH.elapsed()
 		.map(|dur| dur.as_millis())
 		.unwrap_or(0);
 
-	spawn_handle.spawn(
-		"telemetry-worker",
-		telemetry.clone()
-			.for_each(move |event| {
-				// Safe-guard in case we add more events in the future.
-				let sc_telemetry::TelemetryEvent::Connected = event;
-
-				telemetry!(SUBSTRATE_INFO; "system.connected";
-					"name" => name.clone(),
-					"implementation" => impl_name.clone(),
-					"version" => impl_version.clone(),
-					"config" => "",
-					"chain" => chain_name.clone(),
-					"genesis_hash" => ?genesis_hash,
-					"authority" => is_authority,
-					"startup_time" => startup_time,
-					"network_id" => network_id.clone()
-				);
-
-				telemetry_connection_sinks.0.lock().retain(|sink| {
-					sink.unbounded_send(()).is_ok()
-				});
-				ready(())
-			})
-	);
-
-	telemetry
+	let json = serde_json::json! {{
+		"name": name.clone(),
+		"implementation": impl_name.clone(),
+		"version": impl_version.clone(),
+		"config": "",
+		"chain": chain_name.clone(),
+		"genesis_hash": format!("{:?}", genesis_hash),
+		"authority": is_authority,
+		"startup_time": startup_time.to_string(),
+		"network_id": network_id.clone()
+	}};
+
+	let obj = serde_json::from_value(json).expect("it's an object; qed");
```

Comment:
It's just easier to create a json using the macro. But then I want to enforce that the json is an object and not something else.

If I don't do that I will have to make the mapping, use a couple of inserts, works the same really... but if you insist.

Reply:
You create a json and deserialize this. Then you expect on the result. I don't know how this can be easier than just instantiating the object directly.

Please remove this code.

Reply:
Done in b1b554515

Not sure it's much easier though :D have a look